src/dict.c

```diff
@@ -1541,6 +1541,212 @@ dictScanDefrag(dict *d, unsigned long v, dictScanFunction *fn, dictDefragFunctio
+                                                           │
+                                                  ┌────────▼─────────┐
+                                       ┌─────────►│  PrefetchBucket  ├────►────────┐
+                                       │          └────────┬─────────┘            no more tables -> done
+                                       |             bucket|found                  |
+                                       │                   |                       │
+        entry not found - goto next table         ┌────────▼────────┐              │
+                                       └────◄─────┤ PrefetchEntry   |              ▼
+                                    ┌────────────►└────────┬────────┘              │
+                                    |                 Entry│found                  │
+                                    │                      |                       │
+       value not found - goto next entry           ┌───────▼────────┐              |
+                                    └───────◄──────┤ PrefetchValue  |              ▼
+                                                   └───────┬────────┘              │
+                                                      Value│found                  │
+                                                           |                       |
+                                               ┌───────────▼──────────────┐        │
+                                               │    PrefetchValueData     │        ▼
+                                               └───────────┬──────────────┘        │
+                                                           |                       │
+                                                 ┌───────-─▼─────────────┐         │
+                                                 │     PrefetchDone      │◄────────┘
+                                                 └───────────────────────┘
+**********************************************************************************************************************/
+
+typedef struct {
+    PrefetchState state;      /* Current state of the prefetch operation */
+    int ht_idx;               /* Index of the current hash table (0 or 1 for rehashing) */
+    uint64_t bucket_idx;      /* Index of the bucket in the current hash table */
+    uint64_t key_hash;        /* Hash value of the key being prefetched */
+    dictEntry *current_entry; /* Pointer to the current entry being processed */
+} PrefetchInfo;
+
+typedef struct {
+    PrefetchInfo prefetch_info[DictMaxPrefetchSize];
+    size_t current_batch_size; /* Number of keys in the current batch */
+    size_t cur_idx;            /* Index of the current key being prefetched */
+    size_t keys_done;          /* Number of keys that have been processed */
+} PrefetchBatch;
+
+static PrefetchBatch prefetchBatch; /* Global prefetch batch - holds the current batch of keys being prefetched */
```

Comment:
Why is this a global? This could be local just to the prefectBatch function right? I would prefer to avoid globals where possible as it makes the code a bit harder to follow. All the other functions could also include a reference to this global.